File: LeetCodePractice/src/heap/MaximumPerformanceOfaTeam.java
package heap;

import java.util.Arrays;
import java.util.PriorityQueue;

/*
        There are n engineers numbered from 1 to n and two arrays: speed and efficiency, where speed[i] and efficiency[i] represent the speed and efficiency for the i-th engineer respectively. Return the maximum performance of a team composed of at most k engineers, since the answer can be a huge number, return this modulo 10^9 + 7.
        The performance of a team is the sum of their engineers' speeds multiplied by the minimum efficiency among their engineers.

        Example 1:
        Input: n = 6, speed = [2,10,3,1,5,8], efficiency = [5,4,3,9,7,2], k = 2
        Output: 60
        Explanation:
        We have the maximum performance of the team by selecting engineer 2 (with speed=10 and efficiency=4) and engineer 5 (with speed=5 and efficiency=7). That is, performance = (10 + 5) * min(4, 7) = 60.

        Example 2:
        Input: n = 6, speed = [2,10,3,1,5,8], efficiency = [5,4,3,9,7,2], k = 3
        Output: 68
        Explanation:
        This is the same example as the first but k = 3. We can select engineer 1, engineer 2 and engineer 5 to get the maximum performance of the team. That is, performance = (2 + 10 + 5) * min(5, 4, 7) = 68.

        Example 3:
        Input: n = 6, speed = [2,10,3,1,5,8], efficiency = [5,4,3,9,7,2], k = 4
        Output: 72

        Constraints:
        1 <= n <= 10^5
        speed.length == n
        efficiency.length == n
        1 <= speed[i] <= 10^5
        1 <= efficiency[i] <= 10^8
        1 <= k <= n

        similar to MinimumCostToHireKworkers, always sort by common multiply attribute first, but not the sum attribute.
        Then iterate each element common attribute, calculate its best result.

        tag: sort + heap
*/
public class MaximumPerformanceOfaTeam {
    public int maxPerformance(int n, int[] speed, int[] efficiency, int k) {
        final int MOD = (int) (1e9 + 7);
        long res = 0, speedSum = 0;
        Engineer[] engineers = new Engineer[n];
        for(int i = 0; i < n; i++){
            engineers[i] = new Engineer(speed[i], efficiency[i]);
        }
        Arrays.sort(engineers, (a, b) -> b.efficiency - a.efficiency);
        PriorityQueue<Engineer> pq = new PriorityQueue<>((a, b) -> a.speed - b.speed);
        for(Engineer eng : engineers){
            if(pq.size() >= k){
                Engineer slowest = pq.poll();
                speedSum -= slowest.speed;
            }
            speedSum += eng.speed;
            pq.offer(eng);
            res = Math.max(res, speedSum * eng.efficiency);
        }
        return (int)(res % MOD);
    }

    class Engineer {
        int speed, efficiency;
        public Engineer(int speed, int efficiency){
            this.speed = speed;
            this.efficiency = efficiency;
        }
    }
}

File: LeetCodePython/MaximumPerformanceOfaTeam.py
"""
You are given two integers n and k and two integer arrays speed and efficiency both of length n. There are n engineers numbered from 1 to n. speed[i] and efficiency[i] represent the speed and efficiency of the ith engineer respectively.

Choose at most k different engineers out of the n engineers to form a team with the maximum performance.

The performance of a team is the sum of their engineers' speeds multiplied by the minimum efficiency among their engineers.

Return the maximum performance of this team. Since the answer can be a huge number, return it modulo 109 + 7.



Example 1:

Input: n = 6, speed = [2,10,3,1,5,8], efficiency = [5,4,3,9,7,2], k = 2
Output: 60
Explanation:
We have the maximum performance of the team by selecting engineer 2 (with speed=10 and efficiency=4) and engineer 5 (with speed=5 and efficiency=7). That is, performance = (10 + 5) * min(4, 7) = 60.
Example 2:

Input: n = 6, speed = [2,10,3,1,5,8], efficiency = [5,4,3,9,7,2], k = 3
Output: 68
Explanation:
This is the same example as the first but k = 3. We can select engineer 1, engineer 2 and engineer 5 to get the maximum performance of the team. That is, performance = (2 + 10 + 5) * min(5, 4, 7) = 68.
Example 3:

Input: n = 6, speed = [2,10,3,1,5,8], efficiency = [5,4,3,9,7,2], k = 4
Output: 72

Constraints:
1 <= k <= n <= 105
speed.length == n
efficiency.length == n
1 <= speed[i] <= 105
1 <= efficiency[i] <= 108

hint:
1 Keep track of the engineers by their efficiency in decreasing order.
2 Starting from one engineer, to build a team, it suffices to bring K-1 more engineers who have higher efficiencies as well as high speeds.
"""
import heapq
from typing import List


class MaximumPerformanceOfaTeam:
    def maxPerformance(self, n: int, speed: List[int], efficiency: List[int], k: int) -> int:
        mod = 10 ** 9 + 7
        res = speed_sum = 0
        pq = []
        for eff, spd in sorted(zip(efficiency, speed), reverse=1):
            heapq.heappush(pq, spd)
            speed_sum += spd
            if len(pq) > k:
                slowest_spd = heapq.heappop(pq)
                speed_sum -= slowest_spd
            res = max(res, speed_sum * eff)
        return res % mod
